### PR TITLE
Stake Manager - get active validators directly from contract

### PR DESCRIPTION
**Pull request by luka-ciric-ethernal**
Introducing getActiveValidators which get active validators directly from contract for Update Validator Set

### CODE REVIEW

File: contracts/blade/staking/StakeManager.sol [hunk 1]
```diff
@@ -24,6 +26,7 @@ contract StakeManager is IStakeManager, Initializable, Ownable2StepUpgradeable,
     bytes32 public domain;
 
     mapping(address => Validator) public validators;
+    EnumerableSet.AddressSet private activeValidatorsSet;
```
Comment (goran-ethernal) [inline contracts/blade/staking/StakeManager.sol:29]: I dont think introducing a new set adds any value. This just adds more data to the state and increases gas consumed. 

You already have a map containing all the validators, where its IsActive flag is updated when he registers and unstakes.

What you need to do is just use the getValidators function on the client and filter them by the isActive flag, or you can change your getActiveValidators function on the contract to do the filtering, so you dont have to do it on the client.

**Review by goran-ethernal — CHANGES_REQUESTED** [review]
I just realized we have a bug on StakeManager contract. 

When a validator unstakes all, its isActive flag is set to false. But when he tries to stake again, the onlyValidator modifier will stop him from doing so because he is no longer active.

So this is what we need to do:
- change the onlyValidator modifier to check if its blsKey is not nil or empty and that its address is not a zero address. This should be enough to check if the validator is registered.
- when a validator stakes something (amountStaked > 0), update its isActive flag to true.
- remove the newly added set, and just use the validators map like I suggested in my comment.

**Important note**: we need to add a bunch of tests on the client for these edge cases.

**Comment by luka-ciric-ethernal** [comment]
> I just realized we have a bug on StakeManager contract.
> 
> When a validator unstakes all, its isActive flag is set to false. But when he tries to stake again, the onlyValidator modifier will stop him from doing so because he is no longer active.
> 
> So this is what we need to do:
> 
> * change the onlyValidator modifier to check if its blsKey is not nil or empty and that its address is not a zero address. This should be enough to check if the validator is registered.
> * when a validator stakes something (amountStaked > 0), update its isActive flag to true.
> * remove the newly added set, and just use the validators map like I suggested in my comment.
> 
> **Important note**: we need to add a bunch of tests on the client for these edge cases.

When validator unstakes all stake, for it to become validator it needs to be whitelisted & then registered again.

**Comment by goran-ethernal** [comment]
> When validator unstakes all stake, for it to become validator it needs to be whitelisted & then registered again.

That makes sense, but it is not clear in the contract. I remember a case where some client was trying to unstake and exit the validator set, then stake back again.

If we are to follow the Beacon chain on Ethereum, it would mean that we need to retire the validator completely, meaning, to be able to stake again, he needs to register a new key.

But we can discuss this in the future.

I agree, no need to fix the first two points. The third one about the newly added set still stands thiugh.

**Comment by luka-ciric-ethernal** [comment]
We are not going to apply these changes at the moment, so PR will be closed & stored for the future.
